src/Coordination/Changelog.cpp

```diff
@@ -841,18 +844,27 @@ void LogEntryStorage::startCommitLogsPrefetch(uint64_t last_committed_index) con
         if (location_it == logs_location.end())
             throw Exception(ErrorCodes::LOGICAL_ERROR, "Location of log entry with index {} is missing", current_index);
 
-        const auto & [changelog_description, position, size] = location_it->second;
+        const auto & [changelog_description, position, entry_size, size_in_file] = location_it->second;
         if (total_size == 0)
+        {
             current_file_info = &file_infos.emplace_back(changelog_description, position, /* count */ 1);
-        else if (total_size + size > commit_logs_cache.size_threshold)
+            next_position = position + size_in_file;
+        }
+        else if (total_size + entry_size > commit_logs_cache.size_threshold)
             break;
-        else if (changelog_description == current_file_info->file_description)
+        else if (changelog_description == current_file_info->file_description && position == next_position)
```

Comment:
bug was here, I just checked if the files are the same and increased the count of entries to read.
I didn't take into account that we can have old, overwritten entries next to the current entry.
We verify if they are next to each other in file by comparing expected positions.